perceval/utils/persistent_data.py

```diff
@@ -0,0 +1,150 @@
+# license for both exqalibur and Perceval. This exception for the usage of
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import os
+import warnings
+from platformdirs import PlatformDirs
+
+from .metadata import PMetadata
+
+
+class PersistentData:
+    """PersistentData handle perceval persistent data
+    On init, it creates a directory (if it doesn't exist) for storing perceval persistent data
+    Directory depends of the os:
+    * Linux: '/home/my_user/.local/share/perceval-quandela'
+    * Windows: 'C:\\Users\\my_user\\AppData\\Local\\quandela\\perceval-quandela'
+    * Darwin: '/Users/my_user/Library/Application Support/perceval-quandela'
+
+    If the directory cannot be created or read/write in, a warning will inform the user
+    """
+    def __init__(self) -> None:
+        self._directory = PlatformDirs(PMetadata.package_name(), PMetadata.author()).user_data_dir
+        try:
+            self._create_directory()
+        except OSError as exc:
+            warnings.warn(exc)
+            return
+        if not self.is_writable() or not self.is_readable():
+            warnings.warn(UserWarning(f"Cannot read or write in {self._directory}"))
+
+    def is_writable(self) -> bool:
+        """Return if the directory is writable
+
+        :return: True if the directory is writable, False else
+        """
+        return os.access(self._directory, os.W_OK)
+
+    def is_readable(self) -> bool:
+        """Return if the directory is readable
+
+        :return: True if the directory is readable, False else
+        """
+        return os.access(self._directory, os.R_OK)
+
+    def _create_directory(self) -> None:
+        """Create the directory if if doesn't exists
+        """
+        if not os.path.exists(self._directory):
+            os.makedirs(self._directory)
+
+    def get_data_size(self):
+        """Get the directory data size
+
+        :return: directory data size
+        """
+        return sum(os.path.getsize(os.path.join(dirpath, filename))
+                   for dirpath, dirnames, filenames in os.walk(self._directory) for filename in filenames)
+
+    def _get_full_path(self, element_name: str) -> str:
+        """Get the full path of an element supposedly in persistent data directory
+
+        :param element_name: name of the element (with extension)
+        :return: full path of the file
+        """
+        return os.path.join(self._directory, element_name)
+
+    def create_file(self, filename: str):
```

Comment:
Isn't it supposed to be a "write_text_file" method? If not, where are the "read" and "write" method for text files?

Reply:
It's create_binary_file
'x': open for exclusive creation, failing if the file already exists